skbeam/core/correlation.py

```diff
@@ -405,3 +408,358 @@ def auto_corr_scat_factor(lags, beta, relaxation_rate, baseline=1):
+
+    for img in images:
+        s.cur[0] = (1 + s.cur[0]) % num_bufs  # increment buffer
+
+        s.count_level[0] = 1 + s.count_level[0]
+
+        # get the current image time
+        s = s._replace(current_img_time=(s.current_img_time + 1))
+
+        # Put the image into the ring buffer.
+        s.buf[0, s.cur[0] - 1] = (np.ravel(img))[s.pixel_list]
+
+        # Compute the two time correlations between the first level
+        # (undownsampled) frames. two_time and img_per_level in place!
+        _two_time_process(s.buf, s.two_time, s.label_array, num_bufs,
+                          s.num_pixels, s.img_per_level, s.lag_steps,
+                          s.current_img_time, level=0, buf_no=s.cur[0] - 1)
+
+        # time frame for each level
+        time_ind[0].append(s.current_img_time)
+
+        # check whether the number of levels is one, otherwise
+        # continue processing the next level
+        processing = num_levels > 1
+
+        # Compute the correlations for all higher levels.
+        level = 1
+        while processing:
+            if not s.track_level[level]:
+                s.track_level[level] = 1
+                processing = False
+            else:
+                prev = 1 + (s.cur[level - 1] - 2) % num_bufs
+                s.cur[level] = 1 + s.cur[level] % num_bufs
+                s.count_level[level] = 1 + s.count_level[level]
+
+                s.buf[level, s.cur[level] - 1] = (s.buf[level - 1, prev - 1] +
+                                                  s.buf[level - 1,
+                                                  s.cur[level - 1] - 1])/2
+
+                t1_idx = (s.count_level[level] - 1) * 2
+
+                current_img_time = ((time_ind[level - 1])[t1_idx]
+                                    + (time_ind[level - 1])[t1_idx + 1])/2.
+
+                # time frame for each level
+                time_ind[level].append(s.current_img_time)
+
+                # make the track_level zero once that level is processed
+                s.track_level[level] = 0
+
+                # call the _two_time_process function for each multi-tau level
+                # for multi-tau levels greater than one
+                # Again, this is modifying things in place. See comment
+                # on previous call above.
+                _two_time_process(s.buf, s.two_time, s.label_array, num_bufs,
+                                  s.num_pixels, s.img_per_level, s.lag_steps,
+                                  current_img_time, level=level,
+                                  buf_no=s.cur[level]-1)
+                level += 1
+
+                # Checking whether there is next level for processing
+                processing = level < num_levels
+
+    for q in range(np.max(s.label_array)):
+        x0 = (s.two_time)[:, :, q]
+        (s.two_time)[:, :, q] = (np.tril(x0) + np.tril(x0).T -
+                                 np.diag(np.diag(x0)))
+    yield results(s.two_time, s.lag_steps, s)
```

Comment:
This is a different pattern than `lazy_one_time`.  `lazy_one_time` yields after every image. Can you make this function just yield the internal state, `s`, after every image and then provide a helper function to transform the internal state into the results object?